File: app/src/main/java/openfoodfacts/github/scrachx/openfood/views/adapters/ProductImagesSelectionAdapter.java
package openfoodfacts.github.scrachx.openfood.views.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.squareup.picasso.Picasso;

import java.util.List;
import java.util.function.Consumer;

import openfoodfacts.github.scrachx.openfood.R;
import openfoodfacts.github.scrachx.openfood.images.ImageKeyHelper;
import openfoodfacts.github.scrachx.openfood.utils.Utils;

/**
 * Created by prajwalm on 10/09/18.
 */
public class ProductImagesSelectionAdapter extends RecyclerView.Adapter<ProductImagesSelectionAdapter.CustomViewHolder> {
    private final String barcode;
    private final Context context;
    private final List<String> images;
    private final Consumer<Integer> onImageClick;
    int selectedPosition = -1;

    public ProductImagesSelectionAdapter(Context context, List<String> images, String barcode, Consumer<Integer> onImageClick) {
        this.context = context;
        this.images = images;
        this.barcode = barcode;
        this.onImageClick = onImageClick;
    }

    public int getSelectedPosition() {
        return selectedPosition;
    }

    public boolean isSelectionDone() {
        return selectedPosition >= 0;
    }

    @NonNull
    @Override
    public CustomViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        return new CustomViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.image_selectable_item, parent, false));
    }

    public String getSelectedImageName() {
        if (isSelectionDone()) {
            return images.get(selectedPosition);
        }
        return null;
    }

    public String getImageUrl(int position) {
        String imageName = images.get(position);
        return ImageKeyHelper.getImageUrl(barcode, imageName, ImageKeyHelper.IMAGE_EDIT_SIZE_FILE);
    }

    @Override
    public void onBindViewHolder(@NonNull CustomViewHolder holder, int position) {

        String finalUrlString = getImageUrl(position);
        ImageView imageView = holder.productImage;
        ViewGroup viewGroup = holder.parent;
        if (position == selectedPosition) {
            viewGroup.setBackgroundColor(ContextCompat.getColor(context, R.color.blue));
        } else {
            viewGroup.setBackgroundColor(0);
        }
        Utils.picassoBuilder(context).load(finalUrlString).resize(400, 400).centerInside().into(imageView);
    }

    @Override
    public int getItemCount() {
        return images.size();
    }

    class CustomViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        final ViewGroup parent;
        final ImageView productImage;

        public CustomViewHolder(View itemView) {
            super(itemView);
            productImage = itemView.findViewById(R.id.img);
            parent = itemView.findViewById(R.id.parentGroup);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            if (selectedPosition >= 0) {
                notifyItemChanged(selectedPosition);
            }
            int adapterPosition = getAdapterPosition();
            //if the user reclick on the same image -> deselect
            if (adapterPosition == selectedPosition) {
                adapterPosition = -1;
            }
            selectedPosition = adapterPosition;
            if (selectedPosition >= 0) {
                notifyItemChanged(selectedPosition);
            }
            if (onImageClick != null) {
                onImageClick.accept(selectedPosition);
            }
        }
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/adapters/ProductImagesSelectionAdapter.kt
package openfoodfacts.github.scrachx.openfood.features.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.squareup.picasso.Picasso
import openfoodfacts.github.scrachx.openfood.R
import openfoodfacts.github.scrachx.openfood.databinding.ImageSelectableItemBinding
import openfoodfacts.github.scrachx.openfood.images.IMAGE_EDIT_SIZE
import openfoodfacts.github.scrachx.openfood.images.getImageUrl

/**
 * Created by prajwalm on 10/09/18.
 */
class ProductImagesSelectionAdapter(
    private val picasso: Picasso,
    private val images: List<String>,
    private val barcode: String,
    private val onImageClick: ((Int) -> Unit)?,
) : RecyclerView.Adapter<ProductImagesSelectionAdapter.ProductImageSelectionAdapter>() {

    var selectedPosition = -1

    fun isSelectionDone() = selectedPosition >= 0
    fun getSelectedImageName() = if (!isSelectionDone()) null else images[selectedPosition]

    fun getImageUrl(position: Int): String {
        val imageName = images[position]
        return getImageUrl(barcode, imageName, IMAGE_EDIT_SIZE)
    }

    override fun getItemCount() = images.count()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProductImageSelectionAdapter {
        val binding = ImageSelectableItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ProductImageSelectionAdapter(binding)
    }

    override fun onBindViewHolder(holder: ProductImageSelectionAdapter, position: Int) {
        val binding = holder.binding

        val blue = ContextCompat.getColor(binding.root.context, R.color.blue)
        binding.root.setBackgroundColor(if (position == selectedPosition) blue else 0)

        val imageUrl = getImageUrl(position)
        picasso.load(imageUrl)
            .resize(400, 400)
            .centerInside()
            .into(binding.imageView)
    }


    inner class ProductImageSelectionAdapter(
        val binding: ImageSelectableItemBinding,
    ) : RecyclerView.ViewHolder(binding.root) {

        fun onClick() {
            if (selectedPosition >= 0) {
                notifyItemChanged(selectedPosition)
            }

            // If the user reclick on the same image -> deselect
            selectedPosition = bindingAdapterPosition.takeUnless { it == selectedPosition } ?: -1

            if (selectedPosition >= 0) {
                notifyItemChanged(selectedPosition)
            }
            onImageClick?.invoke(selectedPosition)
        }

        init {
            itemView.setOnClickListener { onClick() }
        }
    }
}
